reframe/core/logging.py

```diff
@@ -165,7 +165,40 @@ def _format_time_rfc3339(timestamp, datefmt):
     return re.sub(r'(%)?\:z', tz_rfc3339, time.strftime(datefmt, timestamp))
 
 
-class RFC3339Formatter(logging.Formatter):
+class DynamicFieldFormatter(logging.Formatter):
+    def __init__(self, fmt=None, datefmt=None, style='%'):
+        super().__init__(fmt, datefmt, style)
+        self.__extras = {}
+        fmt = re.findall(r'\%\((check_\S+?)\)s', fmt)
+        for v in set(fmt):
+            if v not in self.__extras:
+                self.__extras[v] = None
```

Comment:
```suggestion
            self.__extras.setdefault(v, None)
```
I think this will also work.

Reply:
These two are not equivalent. The second one will only set the field if it's not present. I don't think that this can ever be the case at this point. You just fill in the `__extras` that you have just created. So @rsarm's original solution is correct.